js/circlepack.js

```diff
@@ -143,29 +155,235 @@ Circlepack.prototype.updateVis = function(){
           : "node node--leaf"
         : "node node--root";
     })
+    .attr("r", d => d.r)
+    .attr("transform", function(d){
+      return `translate(${d.x}, ${d.y})`;
+    })
     .style("fill", function(d){
-      return d.children ? vis.color(d.depth) : null;
+      return d.children ? vis.color(d.depth) : 'white';
     })
     .on("mouseover", function(d) {
-      vis.tooltip.html(`<div id="tooltip-text">${d.data.name}</div>`);
-      vis.tooltip.transition()
-        .duration(300)
-        .style("opacity", 1)
-        // .style("top", `${d.x - d.r - vis.height}px`)
-        // .style("left", `${d.y - d.r}px`)
-    })
-    .on("mousemove", function(d) {
-      vis.tooltip
-        .style("top", d3.event.pageY - 10 + "px")
-        .style("left", d3.event.pageX + 10 + "px");
+      vis.showTooltipOnMouseover(d, this)
     })
     .on("mouseout", function(d) {
-      vis.tooltip.transition()
-        .duration(500)
-        .style("opacity", 0);
-    });
+      vis.hideTooltipOnMouseout(d, this)
+    })
+    .on("click", function(d) {
+      vis.showTooltipOnSelect(d, this);
+      // block circle clicks from propagating to the global click event handler at the top of updateVis
+      d3.event.stopPropagation()
+    })
+    .merge(vis.circles) // update vis with new data
```

Comment:
After the initial load of data for the visualization which happens in the `loadData()` function in `main.js`, if we filter or modify the data we want to show in the visualization, this function tells d3 how to re-render the vis. This is referred to as the "enter-update-exit pattern" in d3, if you're curious to read more.

eg. On the first load of the vis, the `.enter` and `.append` puts all the circle svgs on the page. When the slider changes, we need to filter out the data and redraw the circles accordingly. Without reloading the entire page, d3 does this by setting new attributes on any existing svg elements, and removing svg elements that are no longer represented in the data

Basically, enter = new elements to draw on the screen
update = existing elements on the screen to be modified when the data changes. 
exit = svg elements that should be deleted from the DOM based on the new data

FYI, the merge() function just combines the enter and update "selections" to reduce the amount of code you need to write. ([read here](https://stackoverflow.com/questions/47066905/d3-merge-function))

Reply:
I'm not sure if this would be useful, but you can find a basic intro in the book _Interactive Data Visualization for the Web: 2nd Edition_ by Scott Murray. There's a chapter on updating data and visuals with animated transitions that we learned about in Hanspeter's class (Chapter 9: Updates, Transitions, and Motion)